src/main/scala/backgammon/Hash.scala

```diff
@@ -0,0 +1,894 @@
+package strategygames.backgammon
+
+import strategygames.Player
+
+final class Hash(size: Int) {
+
+  def apply(situation: Situation): PositionHash = {
+    val l = Hash.get(situation, Hash.polyglotTable)
+    if (size <= 8) {
+      Array.tabulate(size)(i => (l >>> ((7 - i) * 8)).toByte)
+    } else {
+      val m = Hash.get(situation, Hash.randomTable)
+      Array.tabulate(size)(i =>
+        if (i < 8) (l >>> ((7 - i) * 8)).toByte
+        else (m >>> ((15 - i) * 8)).toByte
+      )
+    }
+  }
+}
+
+object Hash {
+
+  val size = 3
+
+  class ZobristConstants(start: Int) {
+    def hexToLong(s: String): Long =
+      (java.lang.Long.parseLong(s.substring(start, start + 8), 16) << 32) |
+        java.lang.Long.parseLong(s.substring(start + 8, start + 16), 16)
+    val p1TurnMask                 = hexToLong(ZobristTables.p1TurnMask)
+    val actorMasks                 = ZobristTables.actorMasks.map(hexToLong)
+    val pocketMasks                = ZobristTables.pocketMasks.map(hexToLong)
+  }
+
+  object ZobristConstants {}
+
+  // The following masks are compatible with the Polyglot
+  // opening book format.
+  private val polyglotTable    = new ZobristConstants(0)
+  private lazy val randomTable = new ZobristConstants(16)
+
+  // for backgammon hashInt is just 1 for stone
+  private def pieceIndex(piece: Piece, count: Int) =
+    (piece.role.hashInt * count) * 2 + piece.player.fold(1, 0)
+
+  private def actorIndex(actorCount: (Actor, Int)) =
+    Pos.all.size * pieceIndex(actorCount._1.piece, actorCount._2) + actorCount._1.pos.hashCode
+
+  def get(situation: Situation, table: ZobristConstants): Long = {
+
+    def pocketMask(playershift: Int, count: Int) = {
+      if (0 < count && count <= 15)
+        Option(table.pocketMasks(count + playershift))
+      else None
+    }
+
+    val board = situation.board
+    val hturn = situation.player.fold(table.p1TurnMask, 0L)
+
+    val hactors = board.actors.values.view
+      .map {
+        table.actorMasks compose actorIndex _
+      }
+      .fold(hturn)(_ ^ _)
+
+    board.pocketData.fold(hactors) { data =>
+      Player.all
+        .flatMap { player =>
+          val playershift = player.fold(14, -1)
+          data.pockets(player).roles.groupBy(identity).flatMap { case (_, list) =>
+            pocketMask(playershift, list.size)
+          }
+        }
+        .fold(hactors)(_ ^ _)
+    }
+
+  }
+
+  private val h = new Hash(size)
+
+  def apply(situation: Situation): PositionHash = h.apply(situation)
+
+  def debug(hashes: PositionHash) = hashes.map(_.toInt).sum.toString
+
+}
+
+private object ZobristTables {
+  // to work out the size of this calculate what the max value actorIndex can produce
+  val actorMasks = Array(
```

Comment:
given comments above we can have fewer lines here?
e.g. (14*2 + 1) * 12 + 11 = 359

Reply:
I get:
`24 * ((1*15*2)+1) + 23`